common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -156,26 +157,31 @@ private AppShufflePartitionInfo getOrCreateAppShufflePartitionInfo(
   @VisibleForTesting
   AppShufflePartitionInfo newAppShufflePartitionInfo(
       AppShuffleId appShuffleId,
+      int shuffleSequenceId,
       int reduceId,
       File dataFile,
       File indexFile,
       File metaFile) throws IOException {
-    return new AppShufflePartitionInfo(appShuffleId, reduceId, dataFile,
+    return new AppShufflePartitionInfo(appShuffleId, shuffleSequenceId, reduceId, dataFile,
       new MergeShuffleFile(indexFile), new MergeShuffleFile(metaFile));
   }
 
   @Override
-  public MergedBlockMeta getMergedBlockMeta(String appId, int shuffleId, int reduceId) {
+  public MergedBlockMeta getMergedBlockMeta(
+      String appId,
+      int shuffleId,
+      int shuffleSequenceId,
+      int reduceId) {
     AppShuffleId appShuffleId = new AppShuffleId(appId, shuffleId);
-    File indexFile = getMergedShuffleIndexFile(appShuffleId, reduceId);
+    File indexFile = getMergedShuffleIndexFile(appShuffleId, shuffleSequenceId, reduceId);
```

Comment:
It seems you are changing the fetch side protocols so that you can figure out the `shuffleSequenceId` here to find which files to use. I don't think we should change the fetch side protocols if it's just for this reason.
Instead there should be some logic in the {{RemoteBlockPushResolver}} to know which is the latest shuffle sequence Id

Reply:
How do we get the latest shuffleSequenceId for a shuffle ID? Do you mean to keep track of this information in addition to the existing information we have on shuffle service side? Even then we wouldn't still know whether it is finalized or not right?

Reply:
> Do you mean to keep track of this information in addition to the existing information we have on shuffle service side?

Yes, instead of  modifying the fetch protocols, it should rather be tracked on the server or inferred from existing information (if possible) by the server.

> Even then we wouldn't still know whether it is finalized or not right?

If a partition is not finalized, then a shuffle fetcher will never request to read the shuffle data of that partition. The driver will not have the merge status for that partition.
This comes back to my original point, if the fetch side is always reading the data of latest shuffleSequenceId then these protocols should not me modified.

Reply:
After having offline discussion with @otterc and @mridulm, looks like in the case of a stage re-execution for an indeterminate stage, then all the active jobs which has a chain to this mapStage are rolled back and resubmitted again.
Therefore, we shouldn't hit a case where the old `shuffleSequenceId` data is read for a shuffle. So in terms of correctness, either sending the `shuffleSequenceId` through fetch protocol or keeping the mapping from `<shuffle-id -> <shuffleSequenceId` tracking the latest finalized attempt both should work.

I prefer passing the `shuffleSequenceId` through the `fetch` protocol would keep it simpler but @otterc prefers to not to change the `fetch` protocol as it is not clear when would old `shuffleSequenceId` can be fetched (for that does renaming this variable or adding a comment would help?).
Looking for suggestions if others have any other perspectives, once we agree on one of the solution will go ahead and make that change. cc @Victsm @otterc @mridulm @Ngone51 @tgravescs

Reply:
>  @otterc prefers to not to change the fetch protocol as it is not clear when would old shuffleSequenceId can be fetched (for that does renaming this variable or adding a comment would help?).

@mridulm already mentioned offline that the reducer will never fetch shuffle data of an old `shuffeSequenceId`.  
Already commented here https://github.com/apache/spark/pull/33034#discussion_r656759577 that is is not a matter of renaming and adding comment.